src/Microsoft.Identity.Client/Azure/Exponential.cs

```diff
@@ -0,0 +1,131 @@
+    /// </summary>
+    public class ExponentialBackoff
+    {
+        private readonly int _maxRetries, _delayMilliseconds, _maxDelayMilliseconds;
+        private int _retries, _pow;
+
+        /// <summary>
+        /// Create an instance of an ExponentialBackoff
+        /// </summary>
+        /// <param name="maxRetries">maximum number of retries</param>
+        /// <param name="delayMilliseconds">initial delay in milliseconds</param>
+        /// <param name="maxDelayMilliseconds">maximum delay in milliseconds</param>
+        public ExponentialBackoff(int maxRetries, int delayMilliseconds, int maxDelayMilliseconds)
+        {
+            _maxRetries = maxRetries;
+            _delayMilliseconds = delayMilliseconds;
+            _maxDelayMilliseconds = maxDelayMilliseconds;
+            _retries = 0;
+            _pow = 1;
+        }
+
+        /// <summary>
+        /// DelayAsync will create an exponentially growing delay task
+        /// </summary>
+        /// <returns>a task to be delayed</returns>
+        /// <exception cref="TimeoutException">thrown upon exceeding the max number of retries</exception>
+        public async Task DelayAsync()
+        {
+            if (_retries == _maxRetries) throw new MsalException(ErrorCodes.MaxRetriesExhausted, "Max retry attempts exceeded.");
+            ++_retries;
+            if (_retries < 31) _pow = _pow << 1; // m_pow = Pow(2, m_retries - 1)
+            var delay = Math.Min(_delayMilliseconds * (_pow - 1) / 2, _maxDelayMilliseconds);
+            await Task.Delay(delay).ConfigureAwait(false);
+        }
+    }
+
+    /// <summary>
+    /// TransientManagedIdentityException occurs when a 404, 429 or a 500 series error is encountered.
+    /// 
+    /// see: https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#error-handling
+    /// </summary>
+    public class TransientManagedIdentityException : MsalException
+    {
+        private const string Code = "transient_managed_identity_error";
+        private const string ErrorMessage = "encountered a transient error response from the managed identity service";
+
+        /// <summary>
+        /// Create a TransientManagedIdentityException
+        /// </summary>
+        public TransientManagedIdentityException() : base(Code, ErrorMessage) { }
+
+        // <summary>
+        /// Create a TransientManagedIdentityException with an error message
+        /// </summary>
+        public TransientManagedIdentityException(string errorMessage) : base(Code, errorMessage) { }
+    }
+
+    /// <summary>
+    /// BadManagedIdentityException occurs when a 400 is returned from the managed identity service
+    /// 
+    /// see: https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#error-handling
+    /// </summary>
+    public class BadRequestManagedIdentityException : MsalException
```

Comment:
inherit from MsalServiceException here ? Although it that case this is not coming from Azure AD?

Reply:
In this case, the service would be the managed identity token provider.